### ISSUE TITLE
Critical compilation bug when null analysis enabled

### ISSUE THREAD

**Issue by gayanper** [eclipse-jdt/eclipse.jdt.core#629]
```
import org.springframework.context.ApplicationContextInitializer;
import org.springframework.context.ConfigurableApplicationContext;
import org.springframework.test.context.ContextConfiguration;
@ContextConfiguration(initializers = {AbstractTest.Initializer.class})
public abstract class AbstractTest {

  static class Initializer
      implements ApplicationContextInitializer<ConfigurableApplicationContext> {

    @Override
    public void initialize(ConfigurableApplicationContext configurableApplicationContext) {
        
    }
  }
}
```

When null analysis enabled I get this error

```
Type mismatch: cannot convert from Class<AbstractTest.Initializer> to Class<? extends ApplicationContextInitializer<?>>Java(16777233)
```

**Comment by gayanper** [comment]
@stephan-herrmann I did some investigation and found that when null annotations are enabled the parameterized type resolution for array initializer doesn't contains its super interfaces. That is AbstractTest.Initializer doesn't contains it's super interfaces.

Following is a test snippet I used while investigation

```
	public void testGH629() {
		Map<String, String> options = getCompilerOptions();
		options.put(JavaCore.COMPILER_NONNULL_ANNOTATION_NAME, "test.NonNull");
		options.put(JavaCore.COMPILER_NULLABLE_ANNOTATION_NAME, "test.Nullable");
		// options.put(JavaCore.COMPILER_ANNOTATION_NULL_ANALYSIS, "disable");

		runNegativeTestWithLibs(
				new String[] {
						"Configuration.java",
						"public interface Configuration {\n" +
								"}\n",
						"Init.java",
						"public interface Init<C extends Configuration> {\n" +
								"}\n",
						"Annot.java",
						"public @interface Annot {\n" +
								"    Class<? extends Init<? extends Configuration>>[] inits(); \n" +
								"}\n",
						"App.java",
						"@Annot(inits = {App.MyInit.class})\n" +
								"public class App {\n" +
								"    static class MyInit implements Init<Configuration> {}\n" +
								"}\n"
				},
				options,
				"");
	}
```

**Comment by stephan-herrmann** [comment]
To shed some light on what's happening, have a look at this call stack at the time for unexpected behavior (= when the proposed fix triggers):

```
MemberTypeBinding(ReferenceBinding).implementsInterface(ReferenceBinding, boolean) line: 1346	
MemberTypeBinding(ReferenceBinding).isCompatibleWith0(TypeBinding, Scope) line: 1528	
MemberTypeBinding(ReferenceBinding).isCompatibleWith(TypeBinding, Scope) line: 1450	
MemberTypeBinding(TypeBinding).isCompatibleWith(TypeBinding) line: 679	
MemberTypeBinding(TypeBinding).isTypeArgumentContainedBy(TypeBinding) line: 1343	
ParameterizedTypeBinding.isEquivalentTo(TypeBinding) line: 994	
ParameterizedTypeBinding(ReferenceBinding).isCompatibleWith0(TypeBinding, Scope) line: 1475	
ParameterizedTypeBinding(ReferenceBinding).isCompatibleWith(TypeBinding, Scope) line: 1450	
ParameterizedTypeBinding(TypeBinding).isCompatibleWith(TypeBinding) line: 679	
ArrayInitializer.resolveTypeExpecting(BlockScope, TypeBinding) line: 202	
MemberValuePair.resolveTypeExpecting(BlockScope, TypeBinding) line: 87	
NormalAnnotation(Annotation).resolveType(BlockScope) line: 932	
ASTNode.resolveAnnotations(BlockScope, Annotation[], Binding, boolean) line: 954	
ASTNode.resolveAnnotations(BlockScope, Annotation[], Binding) line: 787	
SourceTypeBinding.initAnnotationTagBits() line: 1657	
SourceTypeBinding.getAnnotationTagBits() line: 1647	
SourceTypeBinding.getNullDefault() line: 2054	
ClassScope.hasDefaultNullnessFor(int, int) line: 1734	
ClassScope.hasDefaultNullnessFor(int, int) line: 1739	
SingleTypeReference(TypeReference).resolveAnnotations(Scope, int) line: 700	
SingleTypeReference(TypeReference).internalResolveType(Scope, int) line: 564	
SingleTypeReference(TypeReference).resolveType(ClassScope, int) line: 634	
SingleTypeReference(TypeReference).resolveTypeArgument(ClassScope, ReferenceBinding, int) line: 652	
ParameterizedSingleTypeReference.internalResolveLeafType(Scope, ReferenceBinding, boolean) line: 263	
ParameterizedSingleTypeReference.internalResolveType(Scope, ReferenceBinding, boolean, int) line: 174	
ParameterizedSingleTypeReference.resolveType(ClassScope, int) line: 395	
ParameterizedSingleTypeReference(TypeReference).resolveType(ClassScope) line: 630	
ParameterizedSingleTypeReference(TypeReference).resolveSuperType(ClassScope) line: 602	
ClassScope.findSupertype(TypeReference) line: 1670	
ClassScope.connectSuperInterfaces() line: 1393	
ClassScope.connectTypeHierarchy() line: 1455	
ClassScope.connectMemberTypes() line: 1143	
ClassScope.connectTypeHierarchy() line: 1464	
CompilationUnitScope.connectTypeHierarchy() line: 394	
LookupEnvironment.completeTypeBindings() line: 521	
```

At the bottom we are connecting bindings in various directions, most interestingly, `connectMemberTypes()` currently processes `MyInit`, and going up in the stack we find ourselves in the middle of `connectSuperInterfaces()` for the same type.

What's specific to null annotations happens going from `SingleTypeReference.resolveAnnotations()` to `ClassScope.hasDefaultNullnessFor()` (which wouldn't happen if that option was disabled). For apparent reasons, finding a `@NonNullByDefault` in any enclosing scope requires those scopes to resolve annotations. Starting from `SourceTypeBinding.getAnnotationTagBits()` we are in code that may perform further resolving *on demand*, which makes this code a bit unpredictable, because by definition we don't know exactly in which state this is executed.

More stuff is triggered because indeed we find the annotation `@Annot(inits = {App.MyInit.class})`, which is where we enter `NormalAnnotation.resolveType()`.

It's when type checking the argument `{App.MyInit.class}` against the declaration `Class<? extends Init<? extends Configuration>>[]` that we find the member type `MyInit` not having assigned its `superInterfaces`. Right, we are in the middle of computing these interfaces down in the stack.

TL;DR: On-demand processing creates a kind of re-entrance where we require information that is still being computed in an active stack frame (and it's not the first time we observed this kind of phenomenon - see https://wiki.eclipse.org/JDT_Core_Programmer_Guide/ECJ#Processing_order !).

So much for the pure analysis.

**Comment by srikanth-sankaran** [comment]
This fix causes https://github.com/eclipse-jdt/eclipse.jdt.core/issues/1007

**Comment by srikanth-sankaran** [comment]
We have a cluster of defects and need a fix that will address all of them. This is what I know at the moment:

 https://github.com/eclipse-jdt/eclipse.jdt.core/issues/854
 https://github.com/eclipse/eclipse.jdt.ls/issues/2386
 https://github.com/eclipse-jdt/eclipse.jdt.core/issues/629
 https://github.com/eclipse-jdt/eclipse.jdt.core/issues/1007
 https://github.com/eclipse-jdt/eclipse.jdt.core/issues/875

@gayanper, for #629 itself, was some analysis done to see if that was a regression and if so what commit led to it ?
Was testGH629() ever working ??

**Comment by gayanper** [comment]
@srikanth-sankaran for #630 I didn't do any analysis since I thought its a issue in jdt.ls

**Comment by stephan-herrmann** [comment]
@srikanth-sankaran I haven't yet looked into all linked issues, but my feeling is, that the common theme might be that evaluating null annotations messes up the order of compilation steps. Can you confirm this?

If that's true, should we switch strategy towards postponing offending tasks to a later/safer point in time? I'm reminded of `ClassScope.deferredBoundChecks` and wonder if that's a good model to copy from / integrate with?

**Comment by srikanth-sankaran** [comment]
> @srikanth-sankaran I haven't yet looked into all linked issues, but my feeling is, that the common theme might be that evaluating null annotations messes up the order of compilation steps. Can you confirm this?
> 
> If that's true, should we switch strategy towards postponing offending tasks to a later/safer point in time? I'm reminded of `ClassScope.deferredBoundChecks` and wonder if that's a good model to copy from / integrate with?

Possibly. I wonder what a visitor based TypeHierarchyConnector phase that produces no side effects other than computing super types would do for this scenario

**Comment by stephan-herrmann** [comment]
> I wonder what a visitor based TypeHierarchyConnector phase that produces no side effects other than computing super types would do for this scenario

Visitor or not, we probably have to sever some existing connections, like when `TypeReference.resolveType` descends into `resolveAnnotations` -> `Scope.hasDefaultNullnessForType`, where side effects happen indeed. It won't be easy to avoid all those side effects, but deep inside we perform some type checking, which indeed is not essential *at that point in time*.

Continue in #1007 ?

**Comment by stephan-herrmann** [comment]
> @stephan-herrmann : since the original fix was reverted, this issue should be reopened? Or should it be part of another one? If yes, which one? I've lost overview now.

I admit the net of issues and PRs has grown a bit complex.

Re-opening to be on the safe side, but should eventually be covered by #1077

**Comment by stephan-herrmann** [comment]
Resolved indeed via #1077
